### PR TITLE
[BuildSystem] Add support for a directory structure tracking node.

### PR DESCRIPTION
**Pull request by ddunbar**
- This is like the existing node which tracks a directory (recursively), but it
   only considers the structural elements of the directory (the files and their
   modes).

### CODE REVIEW

**Comment by ddunbar** [comment]
@swift-ci please smoke test

**Comment by ddunbar** [comment]
@aciidb0mb3r this is the node we talked about...

To use it, create a node like this:
```
nodes:
 "foo/":
    is-directory-structure: true
```
and then create a command with that as an input. It should run every time the structure changes.

Note that the current implementation actually does traverse symbolic links; I view this as a bug (it should be configurable, and off by default), so I would recommend still looking at putting in the SwiftPM logic to add explicit dependencies on the targets of any symbolic links. However, it isn't strictly necessary given the current implementation (and honestly I expect this to be totally sufficient, it is just a shame it will blow up if someone creates a symbolic link to some giant directory).

File: tests/BuildSystem/Build/directory-tree-structure-signatures.llbuild [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+targets:
+  "": ["<all>"]
+
+nodes:
+  "missing-dir/":
+    is-directory-structure: true
+  "file/":
+    is-directory-structure: true
+  "dir/":
+    is-directory-structure: true
```
Comment (aciidb0mb3r) [inline tests/BuildSystem/Build/directory-tree-structure-signatures.llbuild]: Random question: What if I also need `dir/` as a node with is-directory-structure = false?

**Comment by aciidb0mb3r** [comment]
Awesome, thanks!

File: tests/BuildSystem/Build/directory-tree-structure-signatures.llbuild (hunk 1 above)
Comment (ddunbar) [reply tests/BuildSystem/Build/directory-tree-structure-signatures.llbuild]: That may not work out currently... we don't yet have a way to override the target path independent of the name, so unfortunately it has to be mangled this way. I could probably make it work to use "dir//" as a workaround.

**Pull request merged**